### ISSUE TITLE
DbSet<>.Local InvalidOperationException with ListBox.ItemsSource

### ISSUE THREAD

**Issue by ahoischen** [aspnet/EntityFramework#7397]
Using the DbSet<>.Local property as the source for an System.Windows.Controls.ListBox and then adding and removing an instance causes an InvalidOperationException to be thrown. This is not an external issue as this is following your [recommended use case for this property](https://docs.microsoft.com/de-de/ef/core/api/microsoft.entityframeworkcore.dbset-1#Microsoft_EntityFrameworkCore_DbSet_1_Local).
```
System.InvalidOperationException was unhandled
  HResult=-2146233079
  Message=Collection Remove event must specify item position.
  Source=PresentationFramework
  StackTrace:
       at MS.Internal.Data.EnumerableCollectionView.ProcessCollectionChanged(NotifyCollectionChangedEventArgs args)
       at System.Windows.Data.CollectionView.OnCollectionChanged(Object sender, NotifyCollectionChangedEventArgs args)
       at Microsoft.EntityFrameworkCore.ChangeTracking.LocalView`1.OnCollectionChanged(NotifyCollectionChangedEventArgs e)
       at Microsoft.EntityFrameworkCore.ChangeTracking.LocalView`1.StateManagerChangedHandler(InternalEntityEntry entry, EntityState previousState)
       at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.LocalViewListener.StateChanged(InternalEntityEntry entry, EntityState oldState, Boolean fromQuery)
       at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntryNotifier.StateChanged(InternalEntityEntry entry, EntityState oldState, Boolean fromQuery)
       at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntry.SetEntityState(EntityState oldState, EntityState newState, Boolean acceptChanges)
       at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntry.SetEntityState(EntityState entityState, Boolean acceptChanges)
       at Microsoft.EntityFrameworkCore.ChangeTracking.EntityEntry.set_State(EntityState value)
       at Microsoft.EntityFrameworkCore.DbContext.Remove[TEntity](TEntity entity)
       at Microsoft.EntityFrameworkCore.Internal.InternalDbSet`1.Remove(TEntity entity)
       at EntityFramework_bug_local_wpf_listbox.MainWindow.Window_Loaded(Object sender, RoutedEventArgs e) in C:\Users\psych\Documents\git\entityframework-bug-local-wpf-listbox\view\MainWindow.xaml.cs:line 39
       at System.Windows.RoutedEventHandlerInfo.InvokeHandler(Object target, RoutedEventArgs routedEventArgs)
       at System.Windows.EventRoute.InvokeHandlersImpl(Object source, RoutedEventArgs args, Boolean reRaised)
       at System.Windows.UIElement.RaiseEventImpl(DependencyObject sender, RoutedEventArgs args)
       at System.Windows.UIElement.RaiseEvent(RoutedEventArgs e)
       at System.Windows.BroadcastEventHelper.BroadcastEvent(DependencyObject root, RoutedEvent routedEvent)
       at System.Windows.BroadcastEventHelper.BroadcastLoadedEvent(Object root)
       at MS.Internal.LoadedOrUnloadedOperation.DoWork()
       at System.Windows.Media.MediaContext.FireLoadedPendingCallbacks()
       at System.Windows.Media.MediaContext.FireInvokeOnRenderCallbacks()
       at System.Windows.Media.MediaContext.RenderMessageHandlerCore(Object resizedCompositionTarget)
       at System.Windows.Media.MediaContext.RenderMessageHandler(Object resizedCompositionTarget)
       at System.Windows.Media.MediaContext.Resize(ICompositionTarget resizedCompositionTarget)
       at System.Windows.Interop.HwndTarget.OnResize()
       at System.Windows.Interop.HwndTarget.HandleMessage(WindowMessage msg, IntPtr wparam, IntPtr lparam)
       at System.Windows.Interop.HwndSource.HwndTargetFilterMessage(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
       at MS.Win32.HwndWrapper.WndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
       at MS.Win32.HwndSubclass.DispatcherCallbackOperation(Object o)
       at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
       at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
       at System.Windows.Threading.Dispatcher.LegacyInvokeImpl(DispatcherPriority priority, TimeSpan timeout, Delegate method, Object args, Int32 numArgs)
       at MS.Win32.HwndSubclass.SubclassWndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam)
  InnerException: 

```

### Steps to reproduce
I have created an example solution with minimal setup [here](https://github.com/alexander-hoischen/entity-framework-bug-local-wpf-listbox). However I can also quickly summarize the process:
- Create a context
- Set the `ItemsSource` property of a `ListBox` to a local view of one of that context's tables.
- Add an object to that table
- Remove that object

The final step results in the exception above.
Example class:

```c#
public partial class MainWindow : Window
{
    public MainWindow()
    {
        InitializeComponent();
    }

    private void Window_Loaded(object sender, RoutedEventArgs e)
    {
        var context = new FailingModel();
        _listBox.ItemsSource = context.Customers.Local;
        var customer = new Customer();
        context.Customers.Add(customer);

        // The following line throws an InvalidOperationException
        context.Customers.Remove(customer);
    }
}
```

### Further technical details
EF Core version: 1.1.0
Database Provider: Microsoft.EntityFrameworkCore.Sqlite 1.1.0
Operating system: Windows 10 Home x64
IDE: Visual Studio 2015

**Comment by ajcvickers** [comment]
Notes for triage: It looks like the issue here is that the control will only work with an ordered, indexed list. This is not something that our LocalView does. We will need to investigate whether to change LocalView, or instead add a helper for when using it in data binding like this.

**Comment by yajiedesign** [comment]
same problem.Is there any temporary solution?

**Comment by ahoischen** [comment]
The temporary solution for me was to refresh the `Listbox` manually after every change.

**Comment by CKnop** [comment]
Hi @ahoischen,

I have that problem too.
I see you have a working temporary solution by refreshing the Listbox manually.
But I still can't get it to work in my MVVM based solution.
It would be very helpfull if you could update your example solution with the changes you made to get it to work, please?

**Comment by ahoischen** [comment]
@CKnop I can't test the workaround I have right now, but the hack I used was to subscribe to `Local.PropertyChanged` and `Local.CollectionChanged` and create a new Collection as the `ItemsSource` everytime that happened. I *think* it worked, but I'm not sure as I kinda forgot about the project.

**Comment by CKnop** [comment]
@ahoischen Thanks. I can see that Local.CollectionChanged has the correct OldItems value, but the OldStartingIndex is always -1.
Unfortunately OldStartingIndex is readonly (as it should be), so I can't change it to the correct value (as a workaround).
I've successfully changed (in the debugger) the internal value of e._oldStartingIndex = -1 to the correct index of the item that should be removed. And everything worked as it should!
So i think there is a bug in the LocalView code where it needs to set the correct value of e.OldStartingIndex.
I have only tried this when I remove an item. Maybe there is problems with Add too???
The solution I'm using right now is to update the view like this (a bit dirty):
`public List<Employee> Employees { get { return _context.Employees.ToList(); } }`
